[4.x] Hibernate app fails to run (Native Image)

[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"

## Environment Details
* Helidon Version: 4.0.5
* Helidon MP
* JDK version: 21.0.2-graalce
* OS: Mac

----------

## Problem Description
[//]: # "Describe the bug in detail highlighting current behavior vs expected behavior"

Sample JPA/JTA app fails to start with the following exception:
```
% ./target/database-mp 
2023.12.02 23:04:36 INFO io.helidon.logging.jul.JulProvider Thread[#1,main,5,main]: Logging at runtime configured using classpath: /logging.properties
2023.12.02 23:04:36 INFO org.hibernate.jpa.internal.util.LogHelper Thread[#1,main,5,main]: HHH000204: Processing PersistenceUnitInfo [name: pu1]
2023.12.02 23:04:36 INFO org.hibernate.cache.internal.RegionFactoryInitiator Thread[#1,main,5,main]: HHH000026: Second-level cache disabled
Exception in thread "main" org.hibernate.boot.registry.selector.spi.StrategySelectionException: Default-resolver threw exception
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:223)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveDefaultableStrategy(StrategySelectorImpl.java:180)
	at org.hibernate.boot.internal.MetadataBuilderImpl$MetadataBuildingOptionsImpl.<init>(MetadataBuilderImpl.java:746)
	at org.hibernate.boot.internal.MetadataBuilderImpl.<init>(MetadataBuilderImpl.java:140)
	at org.hibernate.boot.MetadataSources.getMetadataBuilder(MetadataSources.java:164)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:277)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:198)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:35)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:102)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilder(HibernatePersistenceProvider.java:169)
	at org.hibernate.jpa.HibernatePersistenceProvider.createContainerEntityManagerFactory(HibernatePersistenceProvider.java:142)
	at io.helidon.integrations.cdi.jpa.PersistenceExtension.produceEntityManagerFactory(PersistenceExtension.java:1409)
	at org.jboss.weld.bootstrap.events.configurator.BeanConfiguratorImpl$CreateCallback.create(BeanConfiguratorImpl.java:372)
	at org.jboss.weld.bootstrap.events.configurator.BeanConfiguratorImpl$ImmutableBean.create(BeanConfiguratorImpl.java:511)
	at org.jboss.weld.contexts.AbstractContext.get(AbstractContext.java:96)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:679)
	at org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:262)
	at org.jboss.weld.bean.builtin.InstanceImpl$InstanceImplIterator.next(InstanceImpl.java:335)
	at io.helidon.integrations.cdi.jpa.JpaExtension.onStartup(JpaExtension.java:2053)
	at java.base@21.0.1/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:85)
	at org.jboss.weld.injection.MethodInvocationStrategy$DefaultMethodInvocationStrategy.invoke(MethodInvocationStrategy.java:144)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)
	at org.jboss.weld.event.ExtensionObserverMethodImpl.sendEvent(ExtensionObserverMethodImpl.java:126)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)
	at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)
	at jakarta.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:142)
	at org.jboss.weld.util.Observers.notify(Observers.java:166)
	at org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285)
	at org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273)
	at org.jboss.weld.event.EventImpl.fire(EventImpl.java:96)
	at io.helidon.microprofile.cdi.HelidonContainerImpl.doStart(HelidonContainerImpl.java:317)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:137)
	at io.helidon.microprofile.cdi.HelidonContainerImpl.start(HelidonContainerImpl.java:256)
	at io.helidon.microprofile.cdi.Main.main(Main.java:80)
	at io.helidon.microprofile.cdi.CdiStartupProvider.start(CdiStartupProvider.java:39)
	at io.helidon.Main.main(Main.java:63)
	at java.base@21.0.1/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: org.hibernate.boot.registry.selector.spi.StrategySelectionException: Could not instantiate named strategy class [org.hibernate.boot.model.relational.ColumnOrderingStrategyStandard]
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:244)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveDefaultableStrategy(StrategySelectorImpl.java:180)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveDefaultableStrategy(StrategySelectorImpl.java:167)
	at org.hibernate.boot.internal.MetadataBuilderImpl$MetadataBuildingOptionsImpl$2.call(MetadataBuilderImpl.java:752)
	at org.hibernate.boot.internal.MetadataBuilderImpl$MetadataBuildingOptionsImpl$2.call(MetadataBuilderImpl.java:749)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:220)
	... 40 more
Caused by: org.hibernate.boot.registry.selector.spi.StrategySelectionException: Could not instantiate named strategy class [org.hibernate.boot.model.relational.ColumnOrderingStrategyStandard]
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.lambda$static$0(StrategySelectorImpl.java:41)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:240)
	... 45 more
Caused by: java.lang.InstantiationException: org.hibernate.boot.model.relational.ColumnOrderingStrategyStandard
	at java.base@21.0.1/java.lang.Class.newInstance(DynamicHub.java:719)
	at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.lambda$static$0(StrategySelectorImpl.java:37)
	... 46 more
Caused by: java.lang.NoSuchMethodException: org.hibernate.boot.model.relational.ColumnOrderingStrategyStandard.<init>()
	at java.base@21.0.1/java.lang.Class.checkMethod(DynamicHub.java:1065)
	at java.base@21.0.1/java.lang.Class.getConstructor0(DynamicHub.java:1228)
	at java.base@21.0.1/java.lang.Class.newInstance(DynamicHub.java:706)
	... 47 more
```

[//]: # "State if the problem is easily reproducible or happens intermittently"
[//]: # "Include stack traces or command outputs"

## Steps to reproduce
- Use Helidon CLI to create brand new database MP application
- `mvn package -Pnative-image`

Run the resulting binary, and you will see the error.

There were some quite major changes done in Hibernate, that broke its usability in native image (at least for us, without doing some major changes).
We will investigate this, if there is an easy solution. 
I am trying to reproduce the problem with latest Helidon (4.0.3), which has an upgraded version of Hibernate. This version no longer supports `none` bytecode provider, and fails at build time of native image with the following:
```
Caused by: org.hibernate.HibernateException: Using the ReflectionOptimizer is not possible when the configured BytecodeProvider is 'none'. Use a different BytecodeProvider
        at org.hibernate.bytecode.internal.none.BytecodeProviderImpl.getReflectionOptimizer(BytecodeProviderImpl.java:48)
```

and if the property is removed:
```
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: An object of type 'net.bytebuddy.implementation.bind.annotation.Argument$Binder' was found in the image heap. This type...
```
Byte buddy should not be reachable in this case at all.

We will try to investigate possible solutions.

--- a/dependencies/pom.xml
+++ b/dependencies/pom.xml
@@ -930,6 +930,11 @@
                 <version>${version.lib.graalvm}</version>
             </dependency>
             <dependency>
+                <groupId>org.graalvm.nativeimage</groupId>
+                <artifactId>svm</artifactId>
+                <version>${version.lib.graalvm}</version>
+            </dependency>
+            <dependency>
                 <groupId>com.graphql-java</groupId>
                 <artifactId>graphql-java</artifactId>
                 <version>${version.lib.graphql-java}</version>

--- a/integrations/cdi/hibernate-cdi/pom.xml
+++ b/integrations/cdi/hibernate-cdi/pom.xml
@@ -74,6 +74,11 @@
             <artifactId>jakarta.transaction-api</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.graalvm.nativeimage</groupId>
+            <artifactId>svm</artifactId>
+            <scope>provided</scope>
+        </dependency>
 
         <!-- Compile-scoped dependencies. -->
         <dependency>

--- a/integrations/cdi/hibernate-cdi/src/main/java/io/helidon/integrations/cdi/hibernate/BytecodeProvider.java
+++ b/integrations/cdi/hibernate-cdi/src/main/java/io/helidon/integrations/cdi/hibernate/BytecodeProvider.java
@@ -1,0 +1,58 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.integrations.cdi.hibernate;
+
+import java.util.Map;
+import java.util.function.Predicate;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.hibernate.bytecode.spi.ReflectionOptimizer;
+import org.hibernate.property.access.spi.PropertyAccess;
+
+/**
+ * Reflection optimizer of the no-op Bytecode provider cannot be disabled
+ * with properties and throw an exception, so it has to return {@code null}.
+ * After Hibernate version 6.5, this substitution is not required anymore.
+ */
+@TargetClass(className = "org.hibernate.bytecode.internal.none.BytecodeProviderImpl",
+        onlyWith = BytecodeProvider.SubstituteOnlyIfPresent.class)
+final class BytecodeProvider {
+
+    private BytecodeProvider() {
+    }
+
+    @Substitute
+    @SuppressWarnings("NullAway")
+    public ReflectionOptimizer getReflectionOptimizer(Class<?> clazz, Map<String, PropertyAccess> propertyAccessMap) {
+        return null;
+    }
+
+    static class SubstituteOnlyIfPresent implements Predicate<String> {
+
+        @Override
+        public boolean test(String type) {
+            try {
+                Class<?> clazz = Class.forName(type, false, getClass().getClassLoader());
+                clazz.getDeclaredMethod("getReflectionOptimizer", Class.class, Map.class);
+                return true;
+            } catch (ClassNotFoundException | NoClassDefFoundError | NoSuchMethodException ex) {
+                return false;
+            }
+        }
+    }
+}

--- a/integrations/cdi/hibernate-cdi/src/main/java/io/helidon/integrations/cdi/hibernate/BytecodeProviderInitiator.java
+++ b/integrations/cdi/hibernate-cdi/src/main/java/io/helidon/integrations/cdi/hibernate/BytecodeProviderInitiator.java
@@ -1,0 +1,54 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.integrations.cdi.hibernate;
+
+import java.util.function.Predicate;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.hibernate.bytecode.spi.BytecodeProvider;
+
+/**
+ * In native image, we force the usage of the no-op bytecode provider so no bytecode
+ * operation happen during runtime.
+ */
+@TargetClass(className = "org.hibernate.bytecode.internal.BytecodeProviderInitiator",
+        onlyWith = BytecodeProviderInitiator.SubstituteOnlyIfPresent.class)
+final class BytecodeProviderInitiator {
+
+    private BytecodeProviderInitiator() {
+    }
+
+    @Substitute
+    public static BytecodeProvider buildBytecodeProvider(String providerName) {
+        return new org.hibernate.bytecode.internal.none.BytecodeProviderImpl();
+    }
+
+    static class SubstituteOnlyIfPresent implements Predicate<String> {
+
+        @Override
+        public boolean test(String type) {
+            try {
+                Class<?> clazz = Class.forName(type, false, getClass().getClassLoader());
+                clazz.getDeclaredMethod("buildBytecodeProvider", String.class);
+                return true;
+            } catch (ClassNotFoundException | NoClassDefFoundError | NoSuchMethodException ex) {
+                return false;
+            }
+        }
+    }
+}

--- a/integrations/cdi/hibernate-cdi/src/main/java/module-info.java
+++ b/integrations/cdi/hibernate-cdi/src/main/java/module-info.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -40,6 +40,7 @@ module io.helidon.integrations.cdi.hibernate {
     requires transitive jakarta.inject;
     requires jakarta.persistence;
     requires transitive jakarta.transaction;
+    requires org.graalvm.nativeimage;
     requires transitive org.hibernate.orm.core;
 
     requires static io.helidon.common.features.api;

--- a/integrations/cdi/hibernate-cdi/src/main/resources/META-INF/helidon/native-image/reflection-config.json
+++ b/integrations/cdi/hibernate-cdi/src/main/resources/META-INF/helidon/native-image/reflection-config.json
@@ -102,6 +102,13 @@
         "org.hibernate.hql.internal.ast.tree.ConstructorNode",
         "org.hibernate.hql.internal.ast.tree.LiteralNode",
         "org.hibernate.hql.internal.ast.tree.BinaryArithmeticOperatorNode",
+        "org.hibernate.query.hql.spi.DotIdentifierConsumer[]",
+        "org.hibernate.query.hql.spi.SqmCreationProcessingState[]",
+        "org.hibernate.query.sqm.spi.ParameterDeclarationContext[]",
+        "org.hibernate.query.sqm.sql.FromClauseIndex[]",
+        "org.hibernate.sql.ast.tree.select.QueryPart[]",
+        "org.hibernate.sql.ast.spi.SqlAstProcessingState[]",
+        "org.hibernate.sql.results.jdbc.spi.JdbcValuesSourceProcessingState[]",
         "antlr.CommonToken"
     ],
     "exclude": [

